helix-term/src/commands.rs

```diff
@@ -2972,19 +2975,31 @@ fn goto_last_diag(cx: &mut Context) {
     doc.set_selection(view.id, selection);
 }
 
-fn goto_next_diag(cx: &mut Context) {
+fn goto_diagnostic_impl(cx: &mut Context, min_severity: Severity, direction: Direction) {
     let (view, doc) = current!(cx.editor);
 
     let cursor_pos = doc
         .selection(view.id)
         .primary()
         .cursor(doc.text().slice(..));
 
-    let diag = doc
+    let mut iter = doc
         .diagnostics()
         .iter()
-        .find(|diag| diag.range.start > cursor_pos)
-        .or_else(|| doc.diagnostics().first());
+        .filter(|diag| diag.severity.unwrap_or(Severity::Hint) >= min_severity);
+
+    let diag = match direction {
+        Direction::Forward => iter
+            .clone()
+            .find(|diag| diag.range.start > cursor_pos)
+            .or_else(|| iter.next()),
+        Direction::Backward => {
+            let mut iter = iter.rev();
+            iter.clone()
+                .find(|diag| diag.range.start < cursor_pos)
+                .or_else(|| iter.next())
+        }
+    };
```

Comment:
```suggestion
    let diag = match direction {
        Direction::Forward => iter
            .find(|diag| diag.range.start > cursor_pos)
            .or_else(|| iter.next()),
        Direction::Backward => {
            let mut iter = iter.rev();
            iter.find(|diag| diag.range.start < cursor_pos)
                .or_else(|| iter.next())
        }
    };
```
Unnecessary clone

Reply:
The clone is necessary as find exhausts the iterator, thus in the suggestion or_else(|| iter.next()) will always return None

Reply:
Should we leave a comment here so we know it shouldn't exhaust the iterator?